src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,228 @@
+    /// </remarks>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed partial class UseConcreteTypeAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1859";
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForMethodReturn = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForMethodReturnMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForParameter = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForParameterMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForLocal = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForLocalMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForField = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForFieldMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            UseConcreteTypeForField,
+            UseConcreteTypeForLocal,
+            UseConcreteTypeForMethodReturn,
+            UseConcreteTypeForParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterSymbolStartAction(c =>
```

Comment:
It is recommended to use the same parameter name `context` for nested registered actions so we don't accidentally register nested actions on outer context.
```suggestion
            context.RegisterSymbolStartAction(context =>
```